src/freenas/usr/local/libexec/ipa_ctl.py

```diff
@@ -0,0 +1,574 @@
+
+DESCRIPTION = (
+    "This program is intended for exclusive use by TrueNAS developers. "
+    "Any use of it outside of the scope of TrueNAS backend operations is "
+    "unsupported and may result in a production outage. "
+    "This program provides support for some basic FreeIPA server related "
+    "operations. "
+    "NOTE: requires the following: "
+    "(1) valid kerberos ticket and configuration for FreeIPA domain, "
+    "(2) valid FreeIPA domain configuration in `/etc/ipa/default.conf`, "
+    "(3) valid FreeIPA domain CA certificate in `/etc/ipa/ca.crt`"
+)
+
+SUPPORTED_SERVICES = ('cifs', 'nfs')
+
+IPA_JOIN = '/sbin/ipa-join'
+IPA_JOIN_CMD_ERR_CODE = {
+    0: 'Success',
+    1: 'Kerberos context initialization failed',
+    2: 'Incorrect usage',
+    3: 'Out of memory',
+    4: 'Invalid service principal name',
+    5: 'No Kerberos credentials cache',
+    6: 'No Kerberos principal and no bind DN and password',
+    7: 'Failed to open keytab',
+    8: 'Failed to create key material',
+    9: 'Setting keytab failed',
+    10: 'Bind password required when using a bind DN',
+    11: 'Failed to add key to keytab',
+    12: 'Failed to close keytab',
+    13: 'Host is already enrolled',
+    14: 'LDAP failure',
+    15: 'Incorrect bulk password',
+    16: 'Host name must be fully-qualified',
+    17: 'RPC fault',
+    18: 'Principal not found in host entry',
+    19: 'Unable to generate Kerberos credentials cache',
+    20: 'Unenrollment result not in RPC response',
+    21: 'Failed to get default Kerberos realm',
+    22: 'Unable to auto-detect fully-qualified hostname'
+}
+
+IPA_GETKEYTAB = '/sbin/ipa-getkeytab'
+IPA_GETKEYTAB_ERR_CODE = {
+    0: 'Success',
+    1: 'Kerberos context initialization failed',
+    2: 'Incorrect usage',
+    3: 'Out of memory',
+    4: 'Invalid service principal name',
+    5: 'No Kerberos credentials cache',
+    6: 'No Kerberos principal and no bind DN and password',
+    7: 'Failed to open keytab',
+    8: 'Failed to create key material',
+    9: 'Setting keytab failed',
+    10: 'Bind password required when using a bind DN',
+    11: 'Failed to add key to keytab',
+    12: 'Failed to close keytab',
+}
+
+DESIRED_ETYPES = (
+    KRB_ETYPE.AES256_CTS_HMAC_SHA1_96.value,
+    KRB_ETYPE.AES128_CTS_HMAC_SHA1_96.value
+)
+
+
+class IpaCtlError(Exception):
+    def __init__(
+        self,
+        op,
+        error_code,
+        rpc_response=None,
+        text=None,
+        error_code_map=None
+    ):
+        self.error_code = error_code
+        self.error_code_str = None
+        self.op = op
+        self.rpc_response = rpc_response
+        self.text = text.strip()
+        self.errmsg = self.__get_errmsg(error_code_map)
+
+    def __get_errmsg(self, error_code_map):
+        if self.rpc_response:
+            return json.dumps(self.rpc_response)
+
+        if self.text:
+            return self.text
+
+        if error_code_map:
+            return error_code_map.get(self.error_code)
```

Comment:
Do we want to handle a `KeyError` or let crash?  (Methinks better to handle, even if we never expect it.)

Reply:
Since it's going through dictionary get, it will default to None if not present, Which I think is OK (though it shouldn't happen).

Reply:
We can fall through though and maybe give more useful error.